### ISSUE TITLE
K.reshape(x, (-1, input_dim)) meaning

### ISSUE THREAD

**Issue by jingweimo** [fchollet/keras#4396]
In reccurent.py, there are some codes like this

    from .. import backend as K
    x = K.reshape(x,(-1,input_dim)) 
    x = K.reshape(x, (-1, timesteps, output_dim))

To see the effect of K.reshape, I type the following codes in Python console:

    from keras import backend as K
    input_dim = 1 # input_dim = x.shape[2], and x has the shape(60000,784,1)
    xnew = K.reshape(x,(-1,input_dim) 

The resultant xnew has the shape 'Shape.0'.
I don't know what is the meaning of this.
And, what the difference of K.reshape() from the conventional reshape function?

Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by pechersky** [comment]
Keras is all about deferred computations on tensors. When you did `xnew.shape`, you got back a `Shape.0` tensor which describes the `xnew` tensor. To actually see the values, you have to do `xnew.shape.eval()`, which will run the backend calculations to get you the shape.

That is also why `K.reshape()` is different from the conventional reshape function -- the conventional function runs on numpy arrays. In Keras, you instead run the tensor functions (using Theano or Tensorflow), and you do not compute the result right away. Instead, all the computations gets stacked, and computed only once at the end.

**Comment by farizrahman4u** [comment]
-1 simply means that the number of dimensions in that axis will be inferred. Since the actual size of a tensor is conserved during a reshape op, this is quite simple. Suppose you have a tensor of shape (3, 2, 3) and you reshape it to (-1, 6), it will get reshaped to (3, 6). 3*2*3 == 3*6.
